snapcraft/file_utils.py

```diff
@@ -110,6 +110,12 @@ def link_or_copy(source, destination, follow_symlinks=False):
         # symlinks.
         os.link(source_path, destination, follow_symlinks=False)
     except OSError:
+        try:
+            # If os.link raised an I/O error, it may have left a file behind.
+            if not os.path.isdir(destination):
+                os.unlink(destination)
+        except FileNotFoundError:
```

Comment:
Two points here:

1. I prefer contextlib for this, @elopio do you prefer try/except?
2. You're already `pass`ing one expected exception, why not stop checking to see if it's a dir and pass on `OSError` as well?